Add support for Sensu etcd replicator

treydock opened the pull request:
# Pull Request Checklist

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Add new type `sensu_etcd_replicator`.

## Related Issue
<!--- Suggest creating an issue first and then referencing it here. -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

Part of #1174 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Sensu Go added support for Etcd replicators and this adds a custom type to manage replicators.

ghoneycutt commented on lib/puppet/provider/sensu_etcd_replicator/sensuctl.rb:
```diff
@@ -0,0 +1,139 @@
+    @property_hash.clear
+  end
+end
+
```
trailing new line

ghoneycutt commented:
Looks great, just need a few small things

- [ ] README entry
- [ ] `tests/` manifest
- [ ] vagrant

treydock commented:
Added vagrant setup that creates two separate backends and creates a replicator on one and verifies the role created on a node is replicator to the other node.  Also updated README.

Added fix to API validator private type to query /version as ran into some odd bugs with /health and both are a way to tell if API is ready to accept requests.

ghoneycutt commented on README.md:
````diff
@@ -576,6 +585,40 @@ sensu::backend::config_hash:
+```
+
+Next configure the Etcd replicator on the backend you wish to push resources from.
+In the following example all defined `Role` resources will be replicated to the backend at the IP address 192.168.52.30.
````
This is a bit confusing with roles and profiles. Not sure what Role means here.

treydock replied:
`Role` is the Sensu Go resource, and copied from Sensu Go documentation.

ghoneycutt commented on README.md:
````diff
@@ -576,6 +585,40 @@ sensu::backend::config_hash:
+
+First configure backend Etcd to listen on an interface besides localhost:
+
+```puppet
````
do we have multiple roles for the sensu backends now such that one is where writes to etcd happen and the other nodes read changes or do all nodes support writes and then sync those between themselves?

treydock replied:
Not sure I follow the question. Based on Sensu Go docs it's designed so you can setup RBAC resources on one cluster and replicate those RBAC resources to another Sensu Go cluster (roles, role bindings, etc).

ghoneycutt replied:
All resources or specific ones?

Curious how replication works and if there is a master in the cluster or if all nodes are equal.

treydock replied:
I don't know for sure, just docs reference RBAC resources but don't mention any limits. I think it's just sending resources from one backend to another when they are not in the same cluster.  I tried to make this PR just a tool to manage the replicator and leave the details and explanations of usage be covered in Sensu Go docs, so far all I've tested is sending RBAC resources per Sensu Go docs. The examples I'm providing and using are all from Sensu Go docs, didn't want to try and deviate too much from that.

ghoneycutt commented on Vagrantfile:
```diff
@@ -73,6 +73,31 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
+    backend.vm.network :forwarded_port, guest: 8081, host: 8086, auto_correct: true
+    backend.vm.provision :shell, :path => "tests/provision_basic_el.sh"
+    backend.vm.provision :shell, :inline => "puppet apply /vagrant/tests/sensu-backend-federated-cluster.pp"
+    backend.vm.provision :shell, :inline => "sensuctl role list"
```
a comment in here like if this works you should see X output at the end of the provisioning step would be helpful.

treydock replied:
I am documenting that in README with latest commit just pushed to support SSL

treydock commented:
Updated the vagrant environment to use SSL for etcd which is the recommended method documented by Sensu Go. We can't use Puppet certs for Etcd because Etcd requires IP SAN and unable to generate certs with Puppet that set IP SANs. The Sensu Go docs document how to generate certs for Etcd and that was the approach I used and put into `tests/etcd-ssl`.

treydock commented:
Running vagrant per README generates this upon second execution of provision:

```
==> sensu-backend-federated1: Running provisioner: shell...
    sensu-backend-federated1: Running: inline script
    sensu-backend-federated1:   Name   Namespace   Rules  
    sensu-backend-federated1:  ────── ─────────── ─────── 
    sensu-backend-federated1:   test   default         1  
```

That role was created on the other vagrant box and replicated to this box via etcd replicator.

ghoneycutt commented on README.md:
````diff
@@ -128,6 +129,18 @@ vagrant up sensu-backend-peer1 sensu-backend-peer2
+vagrant provision sensu-backend-federated1
+```
+
+The `provision` command should output from `sensuctl` the `test` Sensu Go Role that was created on the other backend.
````
Please provide sample output so users can confirm they are getting the desired effect.

treydock replied:
Added output that should be printed.

ghoneycutt commented on tests/README.md:
```diff
@@ -24,6 +24,51 @@ Copy certs from vagrant instance to this repo
+yum install golang-bin
+cd /root
+git clone https://github.com/cloudflare/cfssl.git
+cfssl/
```
i think you mean `cd cfssl`

treydock replied:
typo fixed
